### PR TITLE
Fix sorting/iterator dereferencing

### LINKED ISSUE #4383: Fix PointRef usage as a temporary

### PR DESCRIPTION
**Pull request by abellgithub**
Close #4383

### CODE REVIEW

**Review by connormanning — APPROVED** [review]
The only thing that caught my eye here is the 2 "modes" of the new PointRef, which made me wonder whether it should be a virtual thing so you couldn't accidentally use `m_idx` when there is a `m_view` present.  I think that's probably more trouble than it's worth (and comes with drawbacks like not being able to stack-allocate the thing) so it's better *not* to do it.  So this looks fine to me.

**Comment by abellgithub** [comment]
@connormanning I thought about this kind of thing, but it would break existing code that uses `PointRef` (as you essentially say). This way also avoids virtual calls. I thought about doing some other fancier optimization of `PointRef`, but left it this way for now, unsure that it's worth the time/complexity or that it would make any real difference.

I think I'll add an explicit comment about the two modes.

File: pdal/PointContainer.hpp
[the remark this answers is not in the document]
Comment (abellgithub) [reply pdal/PointContainer.hpp:1]: Done.

**Comment by hobu** [comment]
What should be the changelog note about this update?

**Comment by abellgithub** [comment]
I don't know that one is technically necessary. I guess you could say it's more robust. The behavior should be pretty much identical. We're back to using std::sort instead of std::stable_sort. There could have been algorithms that were broken if they used iterators and temporaries, but we'd probably have heard about it if there was an issue. There are still limitations (always have been), but I don't think they're worth getting into unless someone asks.

**Pull request merged**